### ISSUE TITLE
[평가] 라벨을 -1/0/1 로 쓰면 드러나는 것 — hit_rate 가 중립 38.69% 를 분모에서 뺀다

### ISSUE THREAD

**Issue by devlee328288** [devlee328288/Alpha_Stack#33]
## 한 줄로

**라벨을 `-1/0/1` 로 바꾸는 것 자체는 맞습니다.** 다만 그렇게 명시적으로 쓰기 시작하면, 그동안 한글 문자열 뒤에 가려져 있던 **평가 쪽 네 가지**가 드러납니다. 전부 실측으로 확인했고, 숫자를 같이 올립니다.

> 오준영님이 "하락 중립 상승을 -1 0 1 로 바꿔서 진행하겠다" 고 알려 주셔서 확인하다가 나온 것입니다.
> `evaluation/` 은 강민석님 파트, `models/` 는 오준영님 파트라 **제가 고치지 않고 알려만 드립니다.**

---

## 0. 먼저 — 인코딩은 이미 맞습니다

`evaluation/baseline.py` 가 이미 이렇게 쓰고 있습니다.

```python
UP, DOWN = 1, -1
```

즉 `-1/0/1` 은 충돌이 아니라 **오히려 맞춰지는 쪽**입니다. 👍

우리가 내보내는 CSV 와의 대응은 이렇습니다.

| `features_labels_*.csv` 의 `label` | 모델 |
|---|---|
| `하락` | **-1** |
| `중립` | **0** |
| `상승` | **1** |

순서가 보존돼서(`-1 < 0 < 1` ↔ 하락 < 중립 < 상승) 순서형으로 다뤄도 됩니다.
⚠️ sklearn·LightGBM 은 `classes_` 를 정렬하므로 **`predict_proba` 열 순서가 `하락·중립·상승`** 이 됩니다. 혼동행렬 축도 그 순서로 읽어야 합니다.

---

## 1. 실측 — 우리 라벨 분포

`data/outbox/2026-08-31/small/` 의 실제 파일을 센 값입니다.

| | 하락 | 중립 | 상승 |
|---|---|---|---|
| **코스피200 지수** (2,815행) | 769 (27.32%) | **1,089 (38.69%)** ← 최빈 | 957 (34.00%) |
| 표본 30종목 (56,706행) | **19,324 (34.08%)** ← 최빈 | 18,869 (33.28%) | 18,513 (32.65%) |

**1차 예측 대상인 코스피200 은 중립이 가장 많습니다 (38.69%).** 이게 아래 모든 문제의 뿌리입니다.

---

### ① `hit_rate` 가 중립을 분모에서 뺍니다 — 영향이 가장 큽니다

[`evaluation/metrics.py:112`](../blob/main/evaluation/metrics.py#L112)

```python
mask = a != 0
return float(np.mean(p[mask] == a[mask]))
```

주석에 *"방향이 0(보합)인 시점은 맞고 틀림을 가릴 수 없어 분모에서 뺀다"* 고 적혀 있습니다. **2분류(등락 방향) 맥락에서는 맞는 말**입니다 — 보합은 방향이 없으니까요.

그런데 우리 3분류에서 **중립은 어엿한 예측 대상 클래스**입니다. ±1.0% 밴드 안에 들어온다는 것을 맞히는 것도 실력입니다.

**코스피200 에서 실제로 벌어지는 일:**

```
분모에서 빠짐 : 1,089행 (38.69%)
실제로 재는 행: 1,726행 / 2,815행
```

기준선 `always_up` 을 두 방식으로 재면:

| | 값 |
|---|---|
| `hit_rate` (중립 제외) | **55.45%** |
| 3분류 정확도 (전체) | **34.00%** |
| 차이 | **+21.45%p** ← 기준선이 이만큼 부풀려집니다 |

표본 30종목도 같습니다: 48.93% vs 32.65% (**+16.28%p**).

즉 **3분류 모델을 2분류 지표로 재게 됩니다.** 모델이 중립을 아무리 잘 맞혀도, 못 맞혀도 점수에 나타나지 않습니다.

> ⚠️ 이 동작은 테스트로 못박혀 있습니다 — `tests/test_evaluation.py::test_hit_rate_는_보합을_분모에서_뺀다`. 고치려면 그 테스트도 함께 봐야 합니다.

---

### ② `majority_class` 가 중립을 후보에서 뺍니다

[`evaluation/baseline.py`](../blob/main/evaluation/baseline.py)

```python
n_up   = int(np.sum(y == UP))
n_down = int(np.sum(y == DOWN))
choice = UP if n_up >= n_down else DOWN     # ← 0(중립)은 아예 세지 않는다
```

이름이 *"학습 구간에서 더 많았던 방향"* 인데, **실제로 가장 많았던 것(중립)을 절대 답하지 않습니다.**

**코스피200 실측:**

| | 답하는 것 | 적중률 |
|---|---|---|
| 지금 구현 | 상승 (+1) | 34.00% |
| 진짜 최빈 클래스 | **중립 (0)** | **38.69%** |
| 차이 | | **4.69%p** |

**진짜 최빈 기준선이 4.69%p 더 높습니다.** 기준선이 낮게 잡히면 모델이 실제보다 쉽게 이기는 것처럼 보입니다.

> 표본 30종목에서는 최빈이 하락(-1)이라 **우연히 맞습니다.** 그래서 지금까지 안 보였습니다.

---

### ③ `previous_direction` 이 `0` 을 상승으로 바꿉니다

```python
return np.where(prev == 0, UP, prev)      # 보합 → 상승
```

**코스피200 에서 1,089행(38.69%)** 을 상승으로 바꿔 답합니다. 그중 실제로 상승이었던 것은 **231행**뿐입니다.

3분류에서 "직전 방향이 이어진다"는 기준선이라면 **직전이 중립이면 중립을 답할 수 있어야** 합니다.

---

### ④ 🔴 기준선 **52.64%** 는 우리 3분류의 기준선이 아닙니다

이게 가장 중요할 수 있습니다. 문서 4곳과 팀원 가이드가 이렇게 적고 있습니다.

> 기준선은 **개발구간 52.64%** 이고, 유의미하다고 말하려면 중첩 보정 기준 **57.22%** 를 넘어야 합니다

그런데 그 값을 만드는 [`evaluation/horizon.py:daily_baseline`](../blob/main/evaluation/horizon.py) 은 이렇게 셉니다.

```python
up = sum(1 for r in rows if r["change"] > 0)      # 일간 등락, 2분류
return up / len(rows)
```

**실측으로 재현했습니다:**

```
daily_baseline (일간 change>0, 코스피200 개발구간 2,880일)
  항상 상승 = 52.64%   상승 1,516 · 하락 1,354 · 보합 10
  유의 임계 = 54.99%
```

52.64% 는 정확한 값입니다. **다만 다른 것을 재고 있습니다.**

| | 52.64% 가 재는 것 | 우리 예측 대상 |
|---|---|---|
| 지평 | **1일** | **5거래일** (시가 t+1 → 시가 t+6) |
| 분류 | **2분류** (`change > 0`) | **3분류** (±1.0% 밴드) |
| 기준 | 종가 대비 | 시가 대비 |

**우리 3분류 라벨의 최빈 클래스 기준선은 38.69% 입니다. 13.95%p 차이입니다.**

라벨이 한글 문자열일 때는 이 둘이 뒤섞여도 눈에 안 띄었습니다. `-1/0/1` 로 명시적으로 쓰기 시작하면 **같은 축에 놓게 되고, 그러면 반드시 드러납니다.**

---

### ⑤ (작음) `classify_3` 은 한글 키 dict 를 돌려줍니다

```python
dist = {"상승": 0, "중립": 0, "하락": 0}
```

행 단위 예측이 아니라 **분포를 세는 함수**라 모델과 직접 부딪히지는 않습니다. 다만 `class_balance_ok(dist)` 가 이 dict 를 받으므로, 모델 쪽에서 `-1/0/1` 분포를 넘기려면 옮겨야 합니다.

지금 부르는 곳: `scripts/export_team_dataset.py` · `scripts/measure_horizon.py` · `scripts/measure_stock_horizon.py` (전부 제 것입니다)

---

## 3. 정리 — 무엇이 진짜 문제인가

`-1/0/1` 이 문제를 **만든 것이 아닙니다.** 문제는 원래 있었고, 그동안

- 라벨이 한글 문자열이라 모델 쪽과 평가 쪽이 **아예 안 만났고**
- `evaluation/` 의 지표들이 **2분류(등락 방향)를 전제로** 쓰여 있었는데
- 우리 예측 대상은 **3분류(±1.0% 밴드)** 로 바뀌었기 때문입니다

`-1/0/1` 은 그 둘을 같은 축에 놓습니다. **그래서 드러난 것이지, 그래서 생긴 것이 아닙니다.**

---

### 강민석님 (`evaluation/`)

- [ ] `hit_rate` — 지금 것을 `direction_hit_rate`(2분류 전용)로 이름을 바꾸고, 3분류용으로 **정확도 + macro-F1 + 혼동행렬**을 따로 두는 쪽을 제안합니다. 중립을 빼는 것이 틀린 게 아니라 **다른 질문에 답하는 지표**라서요
- [ ] `majority_class` — 세 클래스를 다 후보에 넣기
- [ ] `previous_direction` — `0` 을 그대로 두기 (3분류에서는 중립도 답이 됩니다)
- [ ] 기준선 표를 **"일간 2분류"** 와 **"5일 3분류"** 두 줄로 나눠 적기

### 오준영님 (`models/`)

- [ ] `predict_proba` 열 순서가 `하락·중립·상승` 인 것 확인 (sklearn 이 `classes_` 를 정렬합니다)
- [ ] 지금 `num_class=3` 으로 두신 것 맞습니다 👍 (`lightgbm.py:53` · `xgboost.py:88`)
- [ ] 기준선과 비교하실 때 **38.69%(3분류 최빈)** 를 쓰시고, 52.64% 는 쓰지 마세요

### 제 쪽 (`docs/`, 이동원)

- [ ] 52.64% 를 인용한 문서 4곳 정정 — `docs/계획서/version2.0/변경사항.md` · `docs/데이터파트/version2.1/명세서.md` · `docs/데이터파트/version2.1/팀원_HuggingFace_가이드.md` · `docs/문제정의/README.md`
- [ ] 5일 3분류의 유의 임계를 다시 계산 (중첩 보정 포함)

⚠️ 위 정정은 **두 분이 방향을 정하신 뒤에** 하겠습니다. 지금 문서만 고치면 코드와 또 어긋납니다.

---

## 5. 재현

```bash
python - <<'EOF'
import pandas as pd, numpy as np, sys
sys.path.insert(0, '.')
from evaluation import baseline
from evaluation.metrics import hit_rate

df = pd.read_csv('data/outbox/2026-08-31/small/features_labels_kospi200_dev.csv')
y = df['label'].dropna().map({'하락': -1, '중립': 0, '상승': 1}).to_numpy(int)

pred = baseline.always_up(y.size)
print('hit_rate (중립 제외):', round(hit_rate(pred, y) * 100, 2), '%')
print('3분류 정확도       :', round((pred == y).mean() * 100, 2), '%')
print('최빈 클래스 기준선  :', round(np.bincount(y + 1).max() / y.size * 100, 2), '%')
EOF
```

```
hit_rate (중립 제외): 55.45 %
3분류 정확도       : 34.0 %
최빈 클래스 기준선  : 38.69 %
```

---

관련: #24 · #26 · PR #32
